tests/webtest_tests.py

```diff
@@ -877,45 +848,168 @@ def test_newest_public_project_and_registrations_show_in_explore_activity(self):
         assert_equal(res.status_code, 200)
 
-    def test_can_reset_password_if_form_success(self):
+    # raise http 400 error if verification_key(OSF) is invalid
+    def test_reset_password_view_raises_400(self):
+        res = self.app.get(self.url_with_invalid_key, expect_errors=True)
+        assert_equal(res.status_code, 400)
+
+    # successfully reset password if osf verification_key(OSF) is valid and form is valid
+    @mock.patch('framework.auth.cas.CasClient.service_validate')
+    def test_can_reset_password_if_form_success(self, mock_service_validate):
+        # load reset password page and submit email
         res = self.app.get(self.url)
         form = res.forms['resetPasswordForm']
         form['password'] = 'newpassword'
         form['password2'] = 'newpassword'
         res = form.submit()
 
-        # password was updated
+        # check request URL is /resetpassword with verification_key(OSF)
+        request_url_path = res.request.path
+        assert_in('resetpassword', request_url_path)
+        assert_in(self.user.verification_key, request_url_path)
+
+        # check verification_key(OSF) is destroyed and a new verification_key(CAS) is in place
+        self.user.reload()
+        self.cas_key = self.user.verification_key
+        assert_not_equal(self.cas_key, self.osf_key)
+
+        # check redirection to CAS login with username and the new verification_key(CAS)
+        assert_equal(res.status_code, 302)
+        location = res.headers.get('Location')
+        assert_true('login?service=' in location)
+        assert_true('username={}'.format(self.user.username) in location)
+        assert_true('verification_key={}'.format(self.user.verification_key) in location)
+
+        # check if password was updated
         self.user.reload()
         assert_true(self.user.check_password('newpassword'))
 
-    @unittest.skip('TODO: Get this working with CAS setup')
+        # check if verification_key(CAS) is destroyed
+        mock_service_validate.return_value = cas.CasResponse(
+            authenticated=True,
+            user=self.user._primary_key,
+            attributes={'accessToken': fake.md5()}
+        )
+        ticket = fake.md5()
+        service_url = 'http://accounts.osf.io/?ticket=' + ticket
+        resp = cas.make_response_from_ticket(ticket, service_url)
+        assert_not_equal(self.user.verification_key, self.cas_key)
+
+    #  logged-in user should be automatically logged out upon before reset password
     def test_reset_password_logs_out_user(self):
-        another_user = AuthUserFactory()
-        # visits reset password link while another user is logged in
-        res = self.app.get(self.url, auth=another_user.auth)
-        assert_equal(res.status_code, 200)
-        # We check if another_user is logged in by checking if
-        # their full name appears on the page (it should be in the navbar).
-        # Yes, this is brittle.
-        assert_not_in(another_user.fullname, res)
```

Comment:
Note for CR: `assert_in('logout?service=', location)` and `assert_not_in('reauth', location)` guarantee that user will be logged out.